BookieScanner

ArvinDevel opened the pull request:
Descriptions of the changes in this PR:

Part one of BookieScanner, includes index file relative stuff.
Master Issue: #<master-issue-number>

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue # or BOOKKEEPER-#>: Description of pull request`
>     `e.g. Issue 123: Description ...`
>     `e.g. BOOKKEEPER-1234: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue # or BOOKKEEPER-#>` in the title with the actual Issue/JIRA number.
> 
> ---

ArvinDevel committed:
BookieScanner v1

ArvinDevel committed:
part of BookieScanner

ArvinDevel committed:
implement indexFile fixing logic

sijie commented:
@ArvinDevel can you rebase your change to my master? so I can review this better.

reddycharan committed:
Dont log ledgermetadata which contains password.

Descriptions of the changes in this PR:

Dont log ledgermetadata contains password used to create
the ledger. Logs are not supposed to reveal such sensitive info.

Author: cguttapalem <cguttapalem@salesforce.com>

Reviewers: Ivan Kelly <ivank@apache.org>, Jia Zhai <None>, Sijie Guo <sijie@apache.org>

This closes #1092 from reddycharan/fixpasswordlog

eolivelli committed:
BP-14 Implementation of WriteFlag.DEFERRED_SYNC on Journal

Implement WriteFlags.DERERRED_SYNC on Bookie side.
In case of DEFERRED_SYNC write Journal will early acknowledge the write, just after flushing the buffer to the OS cache, but before waiting for an fsync.

Author: Enrico Olivelli <eolivelli@apache.org>

Reviewers: Sijie Guo <sijie@apache.org>

This closes #889 from eolivelli/bp14-writeflags-journal

ivankelly committed:
Issue #605: bk_server.conf wrong default logSizeLimit

It's commented out, but the value is invalid. This patch changes it to
match the default in ServerConfiguration.

Author: Ivan Kelly <ivank@apache.org>

Reviewers: Jia Zhai <None>, Sijie Guo <sijie@apache.org>

This closes #1105 from ivankelly/entry-log-size, closes #605

sijie committed:
Run checkstyle plugin at validate phase

Descriptions of the changes in this PR:

Most of the modules run checkstyle plugin at validate phase. However a few modules are still running at `test-compile` phase.
Update the checkstyle plugins in those modules to run at validate phase. So the behavior is consistent across the project.

Author: Sijie Guo <sijie@apache.org>

Reviewers: Jia Zhai <None>

This closes #1098 from sijie/run_checkstyle_at_validate_phase

sijie committed:
BP-27: New BookKeeper CLI

Descriptions of the changes in this PR:

`BookieShell` is the current bookkeeper cli for interacting and operating a bookkeeper cluster. However, this class is getting bigger with more commands added to it. It is facing a few problems for maintenance and extensibility.

- All commands sit in one gaint shell class. It is hard to tell if a command is used for managing a bookie only or if a command is used for managing a cluster.
- Lack of unit tests. This class has very few test coverage. Most of the commands (introduced in early days) don't have a unit test.
- Lack of extensibility. If a new function component (for example, dlog) is introduced, it is a bit hard to extend this CLI to have commands for new function component.

All these problems lead to the proposal here. This proposal is to propose refactoring/redesigning the bookkeeper CLI to allow better managebility for maintenance, better test coverage and better extensibility for new function components.

Master Issue: #1000

Author: Sijie Guo <sijie@apache.org>

Reviewers: Ivan Kelly <ivank@apache.org>, Jia Zhai <None>

This closes #1093 from sijie/BP_27_new_bookkeeper_cli

Commit:
ISSUE #1063: Write keeps refCnt longer

Descriptions of the changes in this PR:

Current code keeps the toSend buffers until client receives
resonses from all wq bookies. From the senders perspective,
It is not required to keep the refcount on it at the PendingAddOp
level, as the ref is taken at bookie client level.

Keeping these buffers longer will increase the memory pressure
on the client.

Signed-off-by: Venkateswararao Jujjuri (JV) <vjujjurisalesforce.com>

Master Issue: #1091

Author: JV Jujjuri <vjujjuri@vjujjuri-ltm2.internal.salesforce.com>

Reviewers: Enrico Olivelli <eolivelli@gmail.com>, Sijie Guo <sijie@apache.org>

This closes #1091 from jvrao/bookkeeper-1063, closes #1063

Commit:
ISSUE #1067: PendingReadOp: recovery, return NoSuchEntry on wQ-aQ+1 errors

In the case of a recovery read, we rely on a NoSuchEntry response on the
first missing entry to determine the final LAC.  As such, we need to
consider an entry to be gone once we see wQ-aQ+1 NoSuchEntry/Ledger
responses from bookies.  Otherwise, a zombie bookie can prevent ledger
recovery from suceeding indefinitely (it'll hit the timeout each time).

There was some preexisting logic from
https://issues.apache.org/jira/browse/BOOKKEEPER-365
b6c1a8bbd7c2d44c2edb59d9938fa073f6f478de,
but it seems to have been too conservative in that it waited for all
responses and required that there were no other errors present.  It
seems now to be unnecessary, and so has been removed.

TestParallelRead.testFailParallelReadMissingEntryImmediately seemed to
rely on the previous logic working outside of recovery, but I believe it
was really meant to test the recovery case, so it has been adjusted.

(bug W-4651456)
Signed-off-by: Samuel Just <sjustsalesforce.com>

Author: Samuel Just <sjust@salesforce.com>

Reviewers: Samuel Just <sjust@salesforce.com>, Ivan Kelly <ivank@apache.org>, Sijie Guo <sijie@apache.org>, Venkateswararao Jujjuri (JV) <None>

This closes #1077 from athanatos/forupstream/wip-4651456-recovery-read, closes #1067

sijie committed:
BP-26 (task 3) : Build dlog shaded jar in bookkeeper way

Descriptions of the changes in this PR:

- discard dlog shaded artifacts and build `distributedlog-core-shaded` jar similar as `bookkeeper-server-shaded`
- discard bookkeeper-server shaded artifact
- move all shaded related tests to `tests/shaded`

Master Issue: #1024

Author: Sijie Guo <sijie@apache.org>

Reviewers: Dave Rusek <dave.rusek@gmail.com>, Jia Zhai <None>, Yiming Zang <yzang2016@gmail.com>

This closes #1095 from sijie/dlog_shade_jar

sijie committed:
BP-26 (task 4): run dlog tests when pull requests modify dlog modules

Descriptions of the changes in this PR:

- enable dlog tests on all post commit CI jobs
- for pull requests, only run dlog tests on travis CI and only when the pull requests modify dlog modules.

Master Issue: #1024

Author: Sijie Guo <sijie@apache.org>

Reviewers: Yiming Zang <yzang2016@gmail.com>, Jia Zhai <None>

This closes #1096 from sijie/enable_dlog_ci

reddycharan committed:
Use loopback network interface for testcases.

Descriptions of the changes in this PR:

- tests would be more reliable irrespective of the environment (and availability of network connection), if the loopback address is used in testsuites.
- unless loopback address is set explicitly, in my env (while on company's VPN),
UpdateCookieCmdTest is failing most of the times.
- currently allowLoopback is set to true in BookKeeperClusterTestCase,
but it doesn't make Bookies to use loopback interface address.
- loopback network interface should be set explicitly as
listening interface to use loopback address.

Note:

1) will create Issue for this depending on the PR feedback.
2) in other places (where setAllowLoopback is set) also loopback interface needs to be set

Author: cguttapalem <cguttapalem@salesforce.com>

Reviewers: Samuel Just <sjust@salesforce.com>, Jia Zhai <None>, Sijie Guo <sijie@apache.org>

This closes #1097 from reddycharan/loopbackaddress

lucperkins committed:
Revamped downloads page

This PR addresses issue #1090 and significantly revamps the downloads page to bring it in line with Apache standards. It also makes some changes to the website's general setup and build process to make that process less brittle.

Author: Luc Perkins <lucperkins@gmail.com>

Reviewers: Jia Zhai <None>, Sijie Guo <sijie@apache.org>

This closes #1104 from lucperkins/lperkins/asf-compliance

sijie committed:
Update release schedule

Descriptions of the changes in this PR:

- update current feature release to 4.7.0 and its release window
- add a section in release guide for instructions to update release schedule

Author: Sijie Guo <sijie@apache.org>

Reviewers: Enrico Olivelli <eolivelli@gmail.com>

This closes #1112 from sijie/update_release_schedule

sijie committed:
Update PR template

Descriptions of the changes in this PR:

- JIRA is not used anymore. Remove it from PR template.

Author: Sijie Guo <sijie@apache.org>

Reviewers: Jia Zhai <None>

This closes #1111 from sijie/update_pr_template

sijie committed:
Change default ledger manager factory from `Flat` to `Hierachical`

Descriptions of the changes in this PR:

the discussion can be found at : http://mail-archives.apache.org/mod_mbox/bookkeeper-dev/201802.mbox/%3CCAO2yDyYdZ8%3D4tViNX1uL0Z67KX78JPD2XBGd7ermjO%3DK%3DscvcQ%40mail.gmail.com%3E

Author: Sijie Guo <sijie@apache.org>

Reviewers: Yiming Zang <yzang2016@gmail.com>, Jia Zhai <None>, Enrico Olivelli <eolivelli@gmail.com>

This closes #1115 from sijie/flip_default_ledger_manager

sijie committed:
Cleanup registration client interface

Descriptions of the changes in this PR:

This change is mainly to remove `zookeeper` reference from metadata interface. The existence of `Optional<ZooKeeper>` is to allow passing an external zookeeper client to bookkeeper, so bookkeeper can reuse that client instance. That is useful for services like pulsar broker, which they can only instantiate a zookeeper client and pass the zookeeper client around to construct bookkeeper client. However, this pollutes the registration client interface, change the method to `Optional<Object>` as an optional context object and let the implementation interpret it.

Author: Sijie Guo <sijie@apache.org>

Reviewers: Enrico Olivelli <eolivelli@gmail.com>, Jia Zhai <None>

This closes #1116 from sijie/clean_registration_client_interface

sijie committed:
Travis CI should not use a hardcoded release version for check-release script

Descriptions of the changes in this PR:

CI is going to be broken when we bumped to 4.8.0.  since travis CI is using a hardcoded version for check-release script.

Author: Sijie Guo <sijie@apache.org>

Reviewers: Ivan Kelly <ivank@apache.org>, Jia Zhai <None>

This closes #1118 from sijie/fix_check_release_script

sijie committed:
Use BookieServer#getLocalAddress() in BookieClientTest

Author: Sijie Guo <sijie@apache.org>

Reviewers: Charan Reddy Guttapalem <reddycharan18@gmail.com>

This closes #1119 from sijie/fix_bookie_client_test

sijie committed:
BP-29: Metadata API module

Descriptions of the changes in this PR:

Related to BP-28 (#1113), this proposal is to propose how we want to organize the metadata modules, to support different metadata storage implementation.

Master Ticket: #1123

Author: Sijie Guo <sijie@apache.org>

Reviewers: Jia Zhai <None>, Sijie Guo <sijie@apache.org>, Venkateswararao Jujjuri (JV) <None>

This closes #1117 from sijie/BP-29-metadata-store-api

ivankelly committed:
Entry count parameters for BenchBookie

So that they can be added in the test, to make the test run faster and
not timeout in CI and ruin an otherwise good test run.

Author: Ivan Kelly <ivank@apache.org>

Reviewers: Enrico Olivelli <eolivelli@gmail.com>

This closes #1136 from ivankelly/bench-flake

ivankelly committed:
Remove spammy log messages in Journal

Looks like they were put in while debugging and not cleaned up.

Author: Ivan Kelly <ivank@apache.org>

Reviewers: Enrico Olivelli <eolivelli@gmail.com>

This closes #1134 from ivankelly/spam-log

jiazhai committed:
[MERGE YAHOO REPO] Only reorder read sequence if any bookie is not available

Descriptions of the changes in this PR:
This is cherry-pick from yahoo repo of branch yahoo-4.3.

original change is:
https://github.com/yahoo/bookkeeper/commit/1ddd14a1
Only reorder read sequence if any bookie is not available

Author: Jia Zhai <zhaijia@apache.org>
Author: Matteo Merli <mmerli@apache.org>

Reviewers: Enrico Olivelli <eolivelli@gmail.com>, Sijie Guo <sijie@apache.org>

This closes #1120 from jiazhai/cherry_picks/i_161

eolivelli committed:
Issue-1143 Error on Java 9 using BookKeeperAdmin API and bookkeeper-server-shaded

Update Maven Shade Plugin to 3.1.0

Relates to issue #1143

Author: Enrico Olivelli <eolivelli@apache.org>

Reviewers: Ivan Kelly <ivank@apache.org>, Sijie Guo <sijie@apache.org>

This closes #1144 from eolivelli/fix/update-shade-plugin

ivankelly committed:
Fix test regression due to change in default ledger manager

TestBenchmark#testReadThroughputLatency() regressed due to the change
in 3ddc5db, where the default ledger manager was changed from flat to
hierarchical.

The fix is to use flat for this test. I've also cleaned up the test a
little, and added some code to the benchmark to make the test run
faster.

Author: Ivan Kelly <ivank@apache.org>

Reviewers: Enrico Olivelli <eolivelli@gmail.com>, Sijie Guo <sijie@apache.org>

This closes #1137 from ivankelly/bench-broke

ivankelly committed:
Require green CI before merge

CI has been broken since the 2018-02-06, due to a non-green patch
being merged. With the recent flake fixes, CI should only go green if
there's a good reason. More over, any remaining problems will get
fixed faster, if people non-green builds start being a pain for
people.

So, this patch blocks merging with the merge script, if all checks are
not green. It can be overriden by adding a comment with the text
"Ignore CI" (case insensitive), on the PR.

Author: Ivan Kelly <ivank@apache.org>

Reviewers: Dave Rusek <dave.rusek@gmail.com>, Jia Zhai <None>, Enrico Olivelli <eolivelli@gmail.com>, Sijie Guo <sijie@apache.org>

This closes #1145 from ivankelly/require-green

ivankelly committed:
Fix shutdown race which left ZK session open

There was a race when shutting down a bookie, where both the main
thread and the shutdown hook thread would try to close the bookie
service. This would result in them racing to set the lifecycle state
and neither would end up cleaning up properly. Specifically, the
starter latch would be counted down, so the main thread would start to
close, then the shutdown hook would run, try to close, and get an
exception when it tried to change the state. Once the all shutdown
hooks end, the process exits, even if main hasn't completed. This
leaves the zookeeper session open.

I've removed the close from main(). The shutdown hook always runs
on a graceful shutdown, so that's the place to do the cleanup.

Author: Ivan Kelly <ivank@apache.org>

Reviewers: Enrico Olivelli <eolivelli@gmail.com>, Sijie Guo <sijie@apache.org>

This closes #1132 from ivankelly/shut-race

ArvinDevel committed:
improve index corrupt impl of bookieScanner
